flips/20201310-storage-fees.md

```diff
@@ -0,0 +1,288 @@
+- the smallest unit of storage purchase is initially set to 10KB but can change as a parameter inside the smart contract.
+- price of storage is 1KB for 1mF.
+- accounts can purchase more storage by calling a method on the StorageFeesContract.
+- one account can purchase storage for another account.
+- accounts can release storage (note that each account can't have less than 100KB, at a cost of 0.1F) and get back the money. This feature should be disabled using a boolean initially to prevent attacks and spam transactions to make money by storage deposits.
+- `FlowServiceAccount.deductAccountCreationFee(account)` (fees contract) calls **StorageFeesContract** and purchases 100KB storage capacity for the account.
+
+### Execution environment changes
+
+The execution environment has three responsibilities:
+
+- keeping the storage used up to date (after every transaction)
+- preventing accounts of going over storage capacity
+
+- giving fvm access to `storage_capacity` and `storage_used`
+    - read access for all accounts to both values
+    - `storage_used` set access to no one (only the execution env changes this)
+    - `storage_capacity` set access only to **StorageFeesContract**
+
+`storage_capacity` and `storage_used` will be stored in the following registers:
+
+```json
+[
+    {
+        "owner": "$address",
+        "controller": "",
+        "key": "storage_capacity"
+    },{
+        "owner": "$address",
+        "controller": "",
+        "key": "storage_used"
+    }
+]
+```
+
+#### Get storage used
+
+Only register values are considered when size is being calculated. Register keys are not part of storage used.
+
+To prevent changing size of the `storage_used` register after `storage_used` is calculated and updated, both `storage_capacity` and `storage_used` registers will be of type uint64. This way both their sizes will be a constant value of 8.
+
+Two possible approaches of calculating the storage used by an address were considered:
+
+- absolute: get all registers for each address (touched in the current `View`) and sum their size to get `storage_used`
+- differential: on the current `View`, keep track of register size changes per address and sum them with the previous `storage_used`.
+
+The decision was made to go with the differential approach, because iterating through all key value pairs in storage for each account touched would be to expensive and would not scale well.
+
+##### Approach 1 - differential
+
+The basic principle of this approach is to update accounts `storage_used` every time a register size changes, by adding the size change to current `storage_used`.
+
+- During `View.Set` and `View.Delete` if storage size changes update `storage_used`.
+- If `storage_used` does not exist on the account add it.
+
+- cons:
+    1. any error (due to code bugs) done in this calculation will be permanent. (e.g. If storage used change is of by 10kB because of a bug in calculation fixing the bug wont correct the storage used)
```

Comment:
```suggestion
    1. any error (due to code bugs) done in this calculation will be permanent. (e.g. If storage used change is of by 10kB because of a bug in calculation fixing the bug will not correct the storage used)
```